grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy

```diff
@@ -102,6 +111,37 @@ class InvoiceController {
         redirect(action: "create", params: [id: invoice.id])
     }
 
+    def generateSalesInvoiceFromShipment = { Invoice invoice ->
+        JSONObject jsonObject = new JSONObject()
+
+        Location currentLocation = Location.get(session.warehouse.id)
+        if (!currentLocation) {
+            throw new IllegalArgumentException("User must be logged into a location to create invoice")
+        }
+
+        StockMovement stockMovement = stockMovementService.getStockMovement(params.id)
+        Shipment shipment = stockMovement.shipment
+        orderService.createOrderFromShipment(shipment, currentLocation)
```

Comment:
Was it required because, in practice, we need the order behind each invoce? Because technically I think it is possible to have invoice associated only to the shipment without order.

Reply:
I think technically it is possible, I would have to modify method invoiceService.createFromShipmentItem() because it uses orderItems. Writing it I was following dev notes from ticket.

Reply:
@MagdaToczek @awalkowiak We should probably create new methods for each object class because I've noticed cases where the data being returned needs to be changed. For example, inbound stock movements aren't represented properly so the data comes back with things like quantityPicked which doesn't make sense in that context. I think this needs a tech huddle, but please feel free to create your own methods to return this data rather than trying to patch an existing method with conditions (i.e. if it's an invoice return this data, if it's an order return this data).